File: src/main/java/at/hannibal2/skyhanni/features/rift/area/mountaintop/quadlinklegacy/TranspositionTable.java
package at.hannibal2.skyhanni.features.rift.area.mountaintop.quadlinklegacy;

import java.util.Arrays;

class TranspositionTable {

    private static final int NUM_ENTRIES = (Config.TABLE_SIZE_MIB * 1024 * 1024) / Long.BYTES;
    private static final int NUM_ROWS = NUM_ENTRIES / 2;

    private static final int INDEX_SIZE = Long.numberOfTrailingZeros(NUM_ROWS);

    private static final long GOLDEN_GAMMA = 0x9e3779b97f4a7c15L;

    private final long[] table = new long[NUM_ENTRIES];

    private final Statistics statistics;

    TranspositionTable(Statistics statistics) {

        this.statistics = statistics;
    }

    void set(long key, int value, int numMoves) {

        statistics.recordTTSet(key);

        int handle = 2 * hash(key);

        long packed1 = table[handle];
        long key1 = Entry.getKey(packed1);
        int value1 = Entry.getValue(packed1);

        long packed2 = table[handle + 1];
        long key2 = Entry.getKey(packed2);
        int value2 = Entry.getValue(packed2);

        if (key == key2) {
            if (value < value2) {
                table[handle + 1] = Entry.toLong(key, value, numMoves);
            }
            return;
        }

        if (key != key1 || value < value1) {

            table[handle] = Entry.toLong(key, value, numMoves);

            if (key != key1 && key1 != 0) {

                if (key2 != 0 && key1 != key2) {
                    statistics.incrementTTCacheEvictions();
                }
                if (key2 == 0 || Entry.getNumMoves(packed1) <= Entry.getNumMoves(packed2)) {
                    table[handle + 1] = packed1;
                }
            }
        }
    }

    int getValueOrDefault(long key, int defaultValue) {

        statistics.recordTTGet(key);

        int handle = 2 * hash(key);

        long packed1 = table[handle];
        long key1 = Entry.getKey(packed1);
        if (key1 == key) {

            statistics.incrementTTHits();
            return Entry.getValue(packed1);
        }

        long packed2 = table[handle + 1];
        long key2 = Entry.getKey(packed2);
        if (key2 == key) {

            statistics.incrementTTHits();
            return Entry.getValue(packed2);
        }
        statistics.incrementTTMisses();
        return defaultValue;
    }

    private int hash(long key) {

        return (int) ((key * GOLDEN_GAMMA) >>> (Long.SIZE - INDEX_SIZE));
    }

    int getEntriesUsed() {

        int count = 0;
        for (long tableEntry : table) {

            if (tableEntry != 0) {
                ++count;
            }
        }
        return count;
    }

    int getCapacity() {

        return NUM_ENTRIES;
    }

    void reset() {

        Arrays.fill(table, 0);
    }

    static class Entry {

        private static final int KEY_SIZE = 48;
        private static final int SCORE_SIZE = 7;
        private static final int NUM_MOVES_SIZE = 6;

        private static final long KEY_MASK = (1L << KEY_SIZE) - 1;
        private static final long SCORE_MASK = (1L << SCORE_SIZE) - 1;
        private static final long NUM_MOVES_MASK = (1L << NUM_MOVES_SIZE) - 1;

        static long toLong(long key, int value, int numMoves) {

            return ((long) numMoves << (KEY_SIZE + SCORE_SIZE)) | (key << SCORE_SIZE) | packScore(value);
        }

        static long getKey(long packed) {

            return (packed >>> SCORE_SIZE) & KEY_MASK;
        }

        static int getValue(long packed) {

            return unpackScore(packed & SCORE_MASK);
        }

        static int getNumMoves(long packed) {

            return (int) ((packed >>> (KEY_SIZE + SCORE_SIZE)) & NUM_MOVES_MASK);
        }

        private static long packScore(int score) {

            return score - Solver.SCORE_MIN + 1L;
        }

        private static int unpackScore(long bits) {

            return (int) bits + Solver.SCORE_MIN - 1;
        }
    }
}

File: src/main/java/at/hannibal2/skyhanni/features/rift/area/mountaintop/quadlinklegacy/TranspositionTable.kt
package at.hannibal2.skyhanni.features.rift.area.mountaintop.quadlinklegacy

import java.util.*

internal class TranspositionTable(private val statistics: Statistics) {
    private val table = LongArray(NUM_BUCKETS)

    fun set(key: Long, value: Int, numMoves: Int) {
        statistics.recordTTSet()

        val handle = 2 * hash(key)

        val packed1 = table[handle]
        val key1 = Entry.getKey(packed1)
        val value1 = Entry.getValue(packed1)

        val packed2 = table[handle + 1]
        val key2 = Entry.getKey(packed2)
        val value2 = Entry.getValue(packed2)

        if (key == key2) {
            if (value < value2) {
                table[handle + 1] = Entry.toLong(key, value, numMoves)
            }
            return
        }

        if (key != key1 || value < value1) {
            table[handle] = Entry.toLong(key, value, numMoves)

            if (key != key1 && key1 != 0L) {
                if (key2 != 0L && key1 != key2) {
                    statistics.incrementTTCacheEvictions()
                }
                if (key2 == 0L || Entry.getNumMoves(packed1) <= Entry.getNumMoves(packed2)) {
                    table[handle + 1] = packed1
                }
            }
        }
    }

    fun getValueOrDefault(key: Long, defaultValue: Int): Int {
        statistics.recordTTGet()

        val handle = 2 * hash(key)

        val packed1 = table[handle]
        val key1 = Entry.getKey(packed1)
        if (key1 == key) {
            statistics.incrementTTHits()
            return Entry.getValue(packed1)
        }

        val packed2 = table[handle + 1]
        val key2 = Entry.getKey(packed2)
        if (key2 == key) {
            statistics.incrementTTHits()
            return Entry.getValue(packed2)
        }
        statistics.incrementTTMisses()
        return defaultValue
    }

    private fun hash(key: Long): Int {
        return ((key * GOLDEN_GAMMA) ushr (Long.SIZE_BITS - INDEX_SIZE)).toInt()
    }

    val entriesUsed: Int
        get() {
            var count = 0
            for (tableEntry in table) {
                if (tableEntry != 0L) {
                    ++count
                }
            }
            return count
        }

    fun reset() {
        Arrays.fill(table, 0L)
    }

    internal object Entry {
        private const val KEY_SIZE = 48
        private const val SCORE_SIZE = 7
        private const val NUM_MOVES_SIZE = 6

        private const val KEY_MASK = (1L shl KEY_SIZE) - 1
        private const val SCORE_MASK = (1L shl SCORE_SIZE) - 1
        private const val NUM_MOVES_MASK = (1L shl NUM_MOVES_SIZE) - 1

        fun toLong(key: Long, value: Int, numMoves: Int): Long {
            return (numMoves.toLong() shl (KEY_SIZE + SCORE_SIZE)) or (key shl SCORE_SIZE) or packScore(value)
        }

        fun getKey(packed: Long): Long {
            return (packed ushr SCORE_SIZE) and KEY_MASK
        }

        fun getValue(packed: Long): Int {
            return unpackScore(packed and SCORE_MASK)
        }

        fun getNumMoves(packed: Long): Int {
            return ((packed ushr (KEY_SIZE + SCORE_SIZE)) and NUM_MOVES_MASK).toInt()
        }

        private fun packScore(score: Int): Long {
            return score - Solver.SCORE_MIN + 1L
        }

        private fun unpackScore(bits: Long): Int {
            return bits.toInt() + Solver.SCORE_MIN - 1
        }
    }

    companion object {
        const val NUM_BUCKETS: Int = (Config.TABLE_SIZE_MIB * 1024 * 1024) / Long.SIZE_BYTES
        private const val NUM_ROWS: Int = NUM_BUCKETS / 2

        private val INDEX_SIZE = NUM_ROWS.toLong().countTrailingZeroBits()

        private const val GOLDEN_GAMMA = -0x61c8864680b583ebL
    }
}
